src/util/retry.rs

```diff
@@ -1,245 +0,0 @@
-
-/// A function that performs and retries the given operation according to a retry policy.
-///
-/// **Caution**: A retry policy without the number of attempts capped by [`MaxAttemptsRetryPolicy`]
-/// decorator will result in infinite retries.
-///
-/// **Example**
-/// ```rust
-/// # use std::time::Duration;
-/// # use vss_client_ng::error::VssError;
-/// # use vss_client_ng::util::retry::{ExponentialBackoffRetryPolicy, retry, RetryPolicy};
-/// #
-/// # async fn operation() -> Result<i32, VssError>  {
-/// # 	tokio::time::sleep(Duration::from_millis(10)).await;
-/// # 	Ok(42)
-/// # }
-/// #
-/// let retry_policy = ExponentialBackoffRetryPolicy::new(Duration::from_millis(100))
-/// 	.with_max_attempts(5)
-/// 	.with_max_total_delay(Duration::from_secs(2))
-/// 	.with_max_jitter(Duration::from_millis(30))
-/// 	.skip_retry_on_error(|e| matches!(e, VssError::InvalidRequestError(..)));
-///
-/// let result = retry(operation, &retry_policy);
-///```
-///
-/// To use a retry policy as a member in a [`Send`] & [`Sync`] safe struct which needs to have known
-/// size at compile time, we can specify its concrete type as follows:
-/// ```
-/// # use std::time::Duration;
-/// # use vss_client_ng::error::VssError;
-/// # use vss_client_ng::util::retry::{ExponentialBackoffRetryPolicy, FilteredRetryPolicy, retry, RetryPolicy};
-///
-/// type VssRetryPolicy = FilteredRetryPolicy<ExponentialBackoffRetryPolicy<VssError>, Box<dyn 'static + Send + Sync + Fn(&VssError) -> bool>>;
-///
-/// struct SomeStruct {
-/// 	retry_policy: VssRetryPolicy,
-/// }
-///
-/// impl SomeStruct {
-/// 	fn new() -> Self {
-/// 		let retry_policy = ExponentialBackoffRetryPolicy::new(Duration::from_millis(100))
-/// 			.skip_retry_on_error(Box::new(|e: &VssError| { matches!( e, VssError::NoSuchKeyError(..)) }) as _);
-/// 		Self { retry_policy }
-/// 	}
-/// }
-/// ```
-pub async fn retry<R, F, Fut, T, E>(mut operation: F, retry_policy: &R) -> Result<T, E>
-where
-	R: RetryPolicy<E = E>,
-	F: FnMut() -> Fut,
-	Fut: Future<Output = Result<T, E>>,
-	E: Error,
-{
-	let mut attempts_made = 0;
-	let mut accumulated_delay = Duration::ZERO;
-	loop {
-		match operation().await {
-			Ok(result) => return Ok(result),
-			Err(err) => {
-				attempts_made += 1;
-				if let Some(delay) = retry_policy.next_delay(&RetryContext {
-					attempts_made,
-					accumulated_delay,
-					error: &err,
-				}) {
-					tokio::time::sleep(delay).await;
-					accumulated_delay += delay;
-				} else {
-					return Err(err);
-				}
-			},
-		}
-	}
-}
-
-/// Provides the logic for how and when to perform retries.
-pub trait RetryPolicy: Sized {
-	/// The error type returned by the `operation` in `retry`.
-	type E: Error;
-
-	/// Returns the duration to wait before trying the next attempt.
-	/// `context` represents the context of a retry operation.
-	///
-	/// If `None` is returned then no further retry attempt is made.
-	fn next_delay(&self, context: &RetryContext<Self::E>) -> Option<Duration>;
```

Comment:
Does `reqwest` implement backoff? It's hard to tell, but there is a placeholder:

https://github.com/seanmonstar/reqwest/blob/a97e1956dd14a79b0207082e327098331519bf2b/src/retry.rs#L49

And they don't seem to use `ExponentialBackoff` provided by the underlying crate.

https://docs.rs/tower/0.5.2/tower/retry/backoff/struct.ExponentialBackoff.html